File: src/leetcode/e26/RemoveDuplicatesFromSortedArray.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.e26;

/**
 * LeetCode 26
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/12/17 10:52
 */
public interface RemoveDuplicatesFromSortedArray {
    int removeDuplicates(int[] nums);

    class Solution1 implements RemoveDuplicatesFromSortedArray {
        @Override
        public int removeDuplicates(int[] nums) {
            if (nums.length == 0) {
                return 0;
            } else {
                int tmp = nums[0];
                int len = 1;
                int index = 1;
                for (int i = 1; i < nums.length; i++) {
                    if (nums[i] != tmp) {
                        tmp = nums[i];
                        nums[index] = tmp;
                        index++;
                        len++;
                    }
                }
                return len;
            }
        }
    }

    class Solution2 implements RemoveDuplicatesFromSortedArray {
        @Override
        public int removeDuplicates(int[] nums) {
            if (nums.length == 1) {
                return 1;
            }
            int slow = 1;
            int fast = 1;
            while (fast < nums.length) {
                if (nums[fast - 1] < nums[fast]) {
                    nums[slow++] = nums[fast];
                }
                fast++;
            }
            return slow;
        }
    }
}

File: src/leetcode/e26/RemoveDuplicatesFromSortedArray.go
package e26

// LeetCode 26
type RemoveDuplicatesFromSortedArray interface {
	removeDuplicates(nums []int) int
}

type Solution1 struct {
}

func (s Solution1) removeDuplicates(nums []int) int {
	idx := 1
	prev := nums[0]
	for i := 1; i < len(nums); i++ {
		if nums[i] != prev {
			nums[idx] = nums[i]
			prev = nums[i]
			idx++
		}
	}
	return idx
}

type Solution2 struct {
}

func (s Solution2) removeDuplicates(nums []int) int {
	idx := 0
	numMap := make(map[int]int)
	for _, num := range nums {
		if _, ok := numMap[num]; !ok {
			numMap[num] = 1
			nums[idx] = num
			idx++
		}
	}
	return idx
}
